src/main/java/com/google/javascript/clutz/ImportRenameMapBuilder.java

```diff
@@ -0,0 +1,128 @@
+      // file is a module, so if it isn't a module, just bail
+      if (moduleId != null) {
+        importRenameMap.putAll(buildImportRenameMap(moduleId, ast));
+      }
+    }
+    return importRenameMap;
+  }
+
+  private boolean isGoogModuleCall(Node statement) {
+    if (!statement.isExprResult()) {
+      return false;
+    }
+
+    Node expression = statement.getFirstChild();
+    return expression.isCall() && expression.getFirstChild().matchesQualifiedName("goog.module");
+  }
+
+  private boolean isGoogRequireAssignment(Node statement) {
+    if (!(statement.isConst() || statement.isVar() || statement.isLet())) {
+      return false;
+    }
+
+    if (statement.getFirstFirstChild() == null) {
+      return false;
+    }
+
+    Node rightHandSide = statement.getFirstFirstChild();
+
+    return rightHandSide.isCall()
+        && rightHandSide.getFirstChild().matchesQualifiedName("goog.require");
+  }
+
+  private boolean isGoogRequireDestructuringAssignment(Node statement) {
+    if (!(statement.isConst() || statement.isVar() || statement.isLet())) {
+      return false;
+    }
+
+    if (!statement.getFirstChild().isDestructuringLhs()) {
+      return false;
+    }
+
+    Node destructuringAssignment = statement.getFirstChild();
+
+    Node rightHandSide = destructuringAssignment.getChildAtIndex(1);
+
+    return rightHandSide.isCall()
+        && rightHandSide.getFirstChild().matchesQualifiedName("goog.require");
+  }
+
+  private String getGoogModuleId(Node astRoot) {
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      if (isGoogModuleCall(statement)) {
+        return statement.getFirstChild().getChildAtIndex(1).getString();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * If we're running in incremental mode, closure gives imported symbols names in the current
+   * module's namespace, instead of the namespace of the module we're importing from. This reads the
+   * import statements to find the module id of the module that's being imported and builds a map
+   * from the names in the current module, to their correct counterparts.
+   */
+  private Map<String, String> buildImportRenameMap(String importingModuleId, Node astRoot) {
+    String abkljeri = astRoot.toStringTree();
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Map<String, String> importRenameMap = new HashMap<>();
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      if (isGoogRequireAssignment(statement)) {
+        String importedModuleId = statement.getFirstFirstChild().getChildAtIndex(1).getString();
+        String variableName = statement.getFirstChild().getString();
+        String symbolIdInCurrentModule =
```

Comment:
would it maybe be a better code organization to split this code into detection and extraction of goog.require in one piece, and then handling the require in a second piece? This here duplicates the string replace and `module$...` stuff in the two branches, and at the same time splits the knowledge about the AST structure of a `goog.require` between this code and the `isGoog...` methods above.

Reply:
I've extracted the string creators, but I don't see a really solid way to make it better beyond that.  The only thing that might be worth extracting is the for loop that grabs the destructured object names, and that'd still need a for loop, so I don't think it's worth it.